src/index/PatternIndex.cpp

```diff
@@ -0,0 +1,834 @@
+}
+
+// _____________________________________________________________________________
+void PatternIndex::createPatterns(VocabularyData* vocabData,
+                                  const std::string& filename_base) {
+  std::string patterns_file_name = filename_base + PATTERNS_SUBJECTS_SUFFIX;
+
+  callCreatePatternsImpl<HasRelationType::SUBJECT, TripleVec::bufreader_type>(
+      _num_bytes_per_predicate, patterns_file_name, &_subjects_pattern_data,
+      _predicate_global_to_local_ids, &_subject_meta_data, _maxNumPatterns,
+      vocabData->langPredLowerBound, vocabData->langPredUpperBound,
+      *vocabData->idTriples);
+
+  _initialized = true;
+}
+
+// _____________________________________________________________________________
+void PatternIndex::createPatternsForObjects(VocabularyData* vocabData,
+                                            const std::string& filename_base) {
+  std::string patterns_file_name = filename_base + PATTERNS_OBJECTS_SUFFIX;
+
+  callCreatePatternsImpl<HasRelationType::OBJECT, TripleVec::bufreader_type>(
+      _num_bytes_per_predicate, patterns_file_name, &_objects_pattern_data,
+      _predicate_global_to_local_ids, &_object_meta_data, _maxNumPatterns,
+      vocabData->langPredLowerBound, vocabData->langPredUpperBound,
+      *vocabData->idTriples);
+
+  _initialized = true;
+}
+
+// _____________________________________________________________________________
+void PatternIndex::createPatternsFromExistingIndex(
+    Id langPredLowerBound, Id langPredUpperBound,
+    IndexMetaDataMmapView& meta_data, ad_utility::File& file,
+    const std::string& filename_base) {
+  std::string patterns_file_name = filename_base + PATTERNS_SUBJECTS_SUFFIX;
+
+  callCreatePatternsImpl<HasRelationType::SUBJECT,
+                         MetaDataIterator<IndexMetaDataMmapView>,
+                         IndexMetaDataMmapView, ad_utility::File>(
+      _num_bytes_per_predicate, patterns_file_name, &_subjects_pattern_data,
+      _predicate_global_to_local_ids, &_subject_meta_data, _maxNumPatterns,
+      langPredLowerBound, langPredUpperBound, meta_data, file);
+
+  _initialized = true;
+}
+
+// _____________________________________________________________________________
+void PatternIndex::createPatternsForObjectsFromExistingIndex(
+    Id langPredLowerBound, Id langPredUpperBound,
+    IndexMetaDataMmapView& meta_data, ad_utility::File& file,
+    const std::string& filename_base) {
+  std::string patterns_file_name = filename_base + PATTERNS_OBJECTS_SUFFIX;
+
+  callCreatePatternsImpl<HasRelationType::OBJECT,
+                         MetaDataIterator<IndexMetaDataMmapView>,
+                         IndexMetaDataMmapView, ad_utility::File>(
+      _num_bytes_per_predicate, patterns_file_name, &_objects_pattern_data,
+      _predicate_global_to_local_ids, &_object_meta_data, _maxNumPatterns,
+      langPredLowerBound, langPredUpperBound, meta_data, file);
+
+  _initialized = true;
```

Comment:
what is the purpose of this`_initialized` variable, is it checked later on?
It also gets true if we are halfway initialized (e.g. Objects but no subjects)